### PR TITLE
opengl3 driver now works on windows including multi window

### PR DESCRIPTION
**Pull request by derammo**
fixed and simplified gl_manager_windows
swap buffers now called for all windows
fixed missing pixel format setting in additional windows
    this makes them work in OpenGL contexts
changed verbose error printing to write once
    this error message happens very frequently while opengl3 is not finished
removed dead code no longer needed after changes
fixed comments that were misinformation
window messages during window creation now handled
    these were previously discarded
    messages now tunnel the required context
changed failure to create opengl3 window on windows to be more fatal
marked a problem with pen code
conditional compilation of vulkan and opengl3 on windows fixed
windows debug builds now show messages on debug console also
rendering driver selection box now shows only compiled drivers
marked some problematic code

thanks to @clayjohn for on-boarding and help/collaboration

### CODE REVIEW

**Comment by derammo** [comment]
I will amend and force push to get rid of the compilation issues.

File: platform/windows/display_server_windows.cpp
Comment (clayjohn) [inline platform/windows/display_server_windows.cpp:3419]: Since this is always compiled out, is it safe to just remove from the PR?

Comment (derammo) [reply platform/windows/display_server_windows.cpp:3419]: I put this here as a suggestion of sorts.  I added this code and then decided not to enable it, because I don't know if you would like it.   If you don't want to do that, then yes it could be removed.

File: servers/rendering/renderer_scene_cull.cpp
Comment (derammo) [reply servers/rendering/renderer_scene_cull.cpp:618]: it does still fail.   ERR_FAIL_COND does print error and then return (void)

I simply split it out into the print error and the return statement, so I could print the error only once.  The only behavior difference is less error printing.

File: platform/windows/detect.py
Comment (derammo) [reply platform/windows/detect.py:272]: ok I implemented this because there are switches in the scons file to turn these on and off, but then they were being ignored on windows specifically, because of the use of the detect script

File: platform/windows/vulkan_context_win.cpp
Comment (derammo) [reply platform/windows/vulkan_context_win.cpp:64]: thanks, fixing and force pushing

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:611]: thanks, fixing and force pushing

File: main/main.cpp
Comment (derammo) [reply main/main.cpp:1324]: thanks, fixing and force pushing

File: drivers/gles3/rasterizer_gles3.cpp
Comment (derammo) [reply drivers/gles3/rasterizer_gles3.cpp:215]: thanks, fixing and force pushing

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:3419]: specifically, the utility is that Windows has API function that operate on "all WS_POPUP windows with owner set to the current window" and so I thought it would be useful to actually set the owner to the main window.  But since we aren't using those functions right now, I think it is up to you if you want it there.

**Comment by derammo** [comment]
I believe that I have resolved the static and compilation issues.  Do I just wait for someone to kick those test builds again and then I get email if they fail?

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:3419]: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-showownedpopups

**Comment by clayjohn** [comment]
> I believe that I have resolved the static and compilation issues. Do I just wait for someone to kick those test builds again and then I get email if they fail?

Pretty much yeah. This just happens on your first PR. Once this is merged tests will run automatically for you.

**Comment by derammo** [comment]
> Pretty much yeah. This just happens on your first PR. Once this is merged tests will run automatically for you.

well at least I have learned that having clang-format on a file saving trigger is pretty much required here....

File: platform/windows/display_server_windows.cpp
Comment (clayjohn) [reply platform/windows/display_server_windows.cpp:3417]: I see, I think it is fine to leave in for now with an update to the comment saying it would be necessary to add if we want to operate on all popups owned by a main window. 

I don't know if the other platforms have equivalent functionality or if it is something that would be useful (this isn't my area of expertise).

**Review by clayjohn — COMMENTED** [review]
Added one more comment, but beyond that this looks good to go!

File: platform/windows/display_server_windows.cpp
Comment (bruvzg) [reply platform/windows/display_server_windows.cpp:3417]: Normally, parent is set only for the exclusive popups (it's handled in `window_set_exclusive`), there's no reason to set it here, and it will most likely break current behavior.

**Review by bruvzg — COMMENTED** [review]
In general, changes seems good.

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:3417]: will terminate with prejudice

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:2238]: What you say is correct.  My intention was that the "main" big switch statement below could process WM_NCCREATE if it wanted to in the future.  Before my changes, NCCREATE and around 6-7 other messages related to initial positioning of the window were being discarded.  I tried to make it so that everything CAN get processed, except for WM_GETMINMAXINFO, because not having a valid window handle below is asking for trouble.      

If you prefer I can just DefWindowProcW(...) instead for sure.   Actually, should I always punt to the user proc if there is one?   I didn't even notice that code at the bottom before.

Something like this used in a gross :) fall through switch/case?
```
			if (user_proc) {
				return CallWindowProcW(user_proc, hWnd, uMsg, wParam, lParam);
			}
			return DefWindowProcW(hWnd, uMsg, wParam, lParam);
```

Comment (derammo) [reply platform/windows/display_server_windows.cpp:2238]: good thing you pointed this out, because I actually didn't resolve window_id at all, I just initialized the window struct but didn't then look for window_id.  So any code below would have failed.  Somehow lost that bit in moving the code around.  New version will be better.

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp]: done

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp]: I reorganized the code and split out the special early handling code into a method, so now the way it is called makes it much clearer that we never continue and always return.  makes it much harder to mess it up later

File: platform/windows/display_server_windows.cpp
Comment (bruvzg) [inline platform/windows/display_server_windows.cpp:611]: This is correct, but the comment is unnecessary.

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:2228]: yes sorry I hadn't found vformat the wrapper :) sorry to waste your time on that

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:611]: Sorry, I would have put in PR comment if I had just filed this one thing.  I was trying to explain why the change but that doesn't belong in the code and could be read as rude.

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:2228]: am I supposed to accept your suggestion (doesn't that create an extra commit?) or should I amend it and force push?

Comment (derammo) [reply platform/windows/display_server_windows.cpp]: Made the change and amended commit.  I wasn't sure if I am supposed to accept suggestions, because that creates more commits?  Decided to go with what is the least work for others.

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp]: Removed comment in amended commit.

File: drivers/gles3/rasterizer_gles3.cpp
Comment (akien-mga) [inline drivers/gles3/rasterizer_gles3.cpp:215]: In `3.x` this is done in `RasterizerGLES3::is_viable()`. Something similar could likely be done here.

File: drivers/gles3/rasterizer_gles3.cpp
Comment (akien-mga) [inline drivers/gles3/rasterizer_gles3.cpp:305]: IMO any comment is a "note" (unless it's an explicit TODO/FIXME/WARNING), so it's redundant to specify it. I think this would be just as fine:
```suggestion
	// Flip content upside down to correct for coordinates.
```

File: main/main.cpp
Comment (akien-mga) [inline main/main.cpp:645]: We use snake case for identifiers.

File: main/main.cpp
Comment (akien-mga) [inline main/main.cpp:1337]: Maybe simpler?
```suggestion
String renderer_hint;

// Start with Vulkan, which will be the default if enabled.
#ifdef VULKAN_ENABLED
	renderer_hint = "vulkan";
#endif

// And OpenGL3 next, or first if Vulkan is disabled.
#ifdef GLES3_ENABLED
	if (!renderer_hint.is_empty()) {
		renderer_hint += ",";
	}
	renderer_hint += "opengl3";
#endif

if (renderer_hint.is_empty()) {
	ERR_PRINT("No rendering driver available.")
}

String default_renderer = renderer_hint.slice(",", 0);
GLOBAL_DEF_RST("rendering/driver/driver_name", default_renderer);
```

File: main/main.cpp
Comment (derammo) [reply main/main.cpp:645]: The only reason I used this gross (in my opinion) code style is because my local declarations didn't compile on most platforms since a goto jumps over this code and that's illegal in gcc.  It isn't my fault that there is a goto to begin with, for the record.  

The extra setting it to "" is intentional to represent the conscious choice of using these values in this case, regardless of what the default is.   I was hoping that together with the comment this would help the next developer decide what should happen in this case (a case that never happens today since building with no driver makes no sense.)

Regarding the snake case, are you blocking this PR for that?

File: platform/windows/display_server_windows.cpp
Comment (akien-mga) [inline platform/windows/display_server_windows.cpp:613]: I know the comment existed already but `// Show the window.` to document what `ShowWindow()` does is not so useful.
Clarifying what `NA` means would be relevant though.

File: drivers/gles3/rasterizer_gles3.cpp
Comment (derammo) [reply drivers/gles3/rasterizer_gles3.cpp:215]: I think this is probably a good idea.  I am worried that your comment will just die here because nobody is going to come back to the PR later.   Should we put that in the code then?   For the record, I am working closely with the current main author of that code, so he is going to fix this (or I will) very shortly anyway, so this is probably academic.

File: platform/windows/display_server_windows.cpp
Comment (akien-mga) [inline platform/windows/display_server_windows.cpp:3432]: Not directly related to your changes but while at it, I'd suggest removing the commented out code above which should obviously not be uncommented (would add two extra parameters to the function call).

File: platform/windows/display_server_windows.cpp
Comment (akien-mga) [inline platform/windows/display_server_windows.cpp:2224]: To match our style guide:
```suggestion
	switch (uMsg) {
		case WM_GETMINMAXINFO: {
			// We receive this during CreateWindowEx and we haven't initialized the window
			// struct, so let Windows figure out the maximized size.
			// Silently forward to user/default.
		} break;
		case WM_NCCREATE: {
			// We tunnel an unowned pointer to our window context (WindowData) through the
			// first possible message (WM_NCCREATE) to fix up our window context collection.
			CREATESTRUCTW *pCreate = (CREATESTRUCTW *)lParam;
			WindowData *pWindowData = reinterpret_cast<WindowData *>(pCreate->lpCreateParams);

			// Fix this up so we can recognize the remaining messages.
			pWindowData->hWnd = hWnd;
		} break;
		default: {
			// Additional messages during window creation should happen after we fixed
			// up the data structures on WM_NCCREATE, but this might change in the future,
			// so report an error here and then we can implement them.
			ERR_PRINT_ONCE(vformat("Unexpected window message 0x%x received for window we cannot recognize in our collection; sequence error.", uMsg));
		} break;
	}
```

File: platform/windows/display_server_windows.cpp
Comment (akien-mga) [inline platform/windows/display_server_windows.cpp:3513]: What does "XXX" mean?

File: platform/windows/gl_manager_windows.cpp
Comment (akien-mga) [inline platform/windows/gl_manager_windows.cpp:135]: Nitpick, but we usually use `err` for error codes.

File: platform/windows/gl_manager_windows.cpp
Comment (akien-mga) [inline platform/windows/gl_manager_windows.cpp:144]:
```suggestion
		ERR_PRINT("Could not attach OpenGL context to newly created window: " + format_error_message(GetLastError()));
```

File: platform/windows/gl_manager_windows.cpp
Comment (akien-mga) [inline platform/windows/gl_manager_windows.cpp:174]:
```suggestion
		ERR_PRINT("Could not detach OpenGL context from newly created window: " + format_error_message(GetLastError()));
```

File: platform/windows/gl_manager_windows.cpp
Comment (akien-mga) [inline platform/windows/gl_manager_windows.cpp:182]:
```suggestion
		ERR_PRINT("Could not attach OpenGL context to newly created window with replaced OpenGL context: " + format_error_message(GetLastError()));
```

File: platform/windows/vulkan_context_win.cpp
Comment (akien-mga) [inline platform/windows/vulkan_context_win.cpp:31]: This is done throughout the Godot codebase, I don't think we want to replicate this comment in a few dozen files.

File: servers/rendering/renderer_scene_cull.cpp
Comment (akien-mga) [inline servers/rendering/renderer_scene_cull.cpp:620]: To be consistent with other cases:
```suggestion
			case RS::INSTANCE_NONE: {
				ERR_PRINT_ONCE("unimplemented base type encountered in renderer scene cull");
				return;
			}
```

File: drivers/gles3/rasterizer_gles3.cpp
Comment (akien-mga) [reply drivers/gles3/rasterizer_gles3.cpp:215]: Yeah then the comment you added is probably fine as is :+1:

File: drivers/gles3/rasterizer_gles3.cpp
Comment (akien-mga) [reply drivers/gles3/rasterizer_gles3.cpp:305]: You can do the changes locally and amend the commit, then force push.

Alternatively if there are lot of suggestions which you want to accept, you can do it with GitHub but then you'll indeed need to squash the extra commit (and possibly drop me as co-author because that's not warranted for style nitpicks :)).

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:3513]: Sorry just like NOTE those are my tags XXX, NOTE, REVISIT

They slip out of my hands automatically and should not be in your code.

File: main/main.cpp
Comment (akien-mga) [reply main/main.cpp:645]: > The only reason I used this gross (in my opinion) code style is because my local declarations didn't compile on most platforms since a goto jumps over this code and that's illegal in gcc. It isn't my fault that there is a goto to begin with, for the record.

Ah I see, that doesn't compile indeed.

> The extra setting it to "" is intentional to represent the conscious choice of using these values in this case, regardless of what the default is. I was hoping that together with the comment this would help the next developer decide what should happen in this case (a case that never happens today since building with no driver makes no sense.)

There are 6839 String declarations in the codebase. Them defaulting to an empty String *is* already a conscious decision and it's redundant to specify it again.

> Regarding the snake case, are you blocking this PR for that?

Yes, we have a code style and it's important to respect it. We all work together on this codebase and the code needs to be familiar to all contributors.

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:3513]: To answer your question XXX in my code bases is "do not ship if build includes this tag", but yeah that was by mistake.

File: platform/windows/display_server_windows.cpp
Comment (akien-mga) [reply platform/windows/display_server_windows.cpp:2199]: BTW I'm aware that some pre-existing comments may often have a different style. We're trying to enforce a consistent style for *new* comments at least so that progressively the surrounding code makes it obvious what the preferred style is. See https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#comment-style-guide

File: main/main.cpp
Comment (derammo) [reply main/main.cpp:1337]: is that going to compile?   are you allowed to goto over that stack declaration?

File: platform/windows/vulkan_context_win.cpp
Comment (derammo) [reply platform/windows/vulkan_context_win.cpp:31]: I read in the style guide to put a comment into the code explaining why you are changing something.  Then I was told the opposite in the developer chat channels.   Now I know.

File: main/main.cpp
Comment (derammo) [reply main/main.cpp:645]: > There are 6839 String declarations in the codebase. Them defaulting to an empty String _is_ already a conscious decision and it's redundant to specify it again.

That's an opinion and I disagree.  Since it is your code base of course you can do as you wish.

I was trying to communicate with the next developer down the line and since the static analyzer in the compiler removes that second initialization anyway, I didn't see any downside.   I will of course change it to suit your preference, but I won't pretend like I don't know what I am doing.

File: main/main.cpp
Comment (derammo) [reply main/main.cpp:1337]: What makes this no longer fun is that what I provided is much better than what was there (failure to consider the conditional compilation flags at all.)   Because I tried to fix that, I am now stuck with all this nitpicking.   I guess if I had made separate PRs then I could have just withdrawn the ones that aren't worth it, but I was really trying to get out the important material changes that are hurting opengl3 development that is happening right now.   So no, I don't care even a tiny bit about how nice this temporary hack of hard coded list is.  It's just not worth my time to work on that.  And I understand now that this means this code base and I aren't going to get along. :)    And that's fine too.

Comment (derammo) [reply main/main.cpp:1337]: @akien-mga I will finish this PR (today) so that we can put this all behind us :).  Please be specific on what you need to accept it.

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:613]: It is a windows API constant.   I believe anyone working in this area of the code will look it up in the MS references if they need to?

File: platform/windows/gl_manager_windows.cpp
Comment (derammo) [reply platform/windows/gl_manager_windows.cpp:135]: I agree, that is a nitpick.

File: main/main.cpp
Comment (derammo) [reply main/main.cpp:1337]: I put in your code, even though I don't believe String+= and get_slice are a simpler solution than conditionally compiling a single string constant to different values.  I think the best would actually be to but the #ifdef tree at the top of the function and just set a single const char* so it is static data in a data segment.   That would be far away from where it is used, which I didn't like originally.

Comment (derammo) [reply main/main.cpp:1337]: Correction: it is also static data in the code I had that you didn't like, it just gets converted to a String object, which makes sense I think in the code base.

File: main/main.cpp
Comment (derammo) [reply main/main.cpp:645]: snakes have been added

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:613]: added a link to the API docs

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:2199]: > Our comment style is to use normal sentence typography - start with a capital letter and end with a dot. Also I think it should be "window messages" and not the other way around?

we are both wrong :)    it should be "a single window message" which I have now put instead

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:3430]: I figured reduzio left it there for a reason, maybe to consider using it again later?  I don't have sufficient ownership to remove their stuff.

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp:3511]: removed

File: platform/windows/gl_manager_windows.cpp
Comment (derammo) [reply platform/windows/gl_manager_windows.cpp:135]: changed

File: platform/windows/vulkan_context_win.cpp
Comment (derammo) [reply platform/windows/vulkan_context_win.cpp:31]: removed

File: platform/windows/gl_manager_windows.cpp
Comment (derammo) [reply platform/windows/gl_manager_windows.cpp]: capitalized

**Comment by derammo** [comment]
@akien-mga I believe I corrected all your blockers.

**Comment by akien-mga** [comment]
Thanks, I'll re-review tomorrow. I apologize if my review comments were frustrating, this was not my intention. Godot is a big codebase with a lot of written and unwritten conventions which make a project with 1600+ co-authors possible *and* maintainable by the few dozens who stick around long term. I can understand that upholding these requirements on things which are not critical to what the PR does can be annoying.

**Comment by derammo** [comment]
@akien-mga I thank you for your kind words.  Yes, you did make me decide to quit since I don't want to spend my time on minutiae, as I am old and grumpy.  I only started last week but I think I will give my relationship with godot another chance based on your clarification.

File: platform/windows/display_server_windows.cpp
Comment (derammo) [reply platform/windows/display_server_windows.cpp]: added URL.  MS reference URLs are usually good for several years

**Review by akien-mga — APPROVED** [review]
Looks good to me. I'll let @bruvzg review for the Windows bits and @clayjohn for OpenGL.

**Review by clayjohn — APPROVED** [review]
Looks good from my end!

**Comment by akien-mga** [comment]
Thanks!

**Comment by Knand1861** [comment]
With godot.4.11a problem is back.

**Comment by Calinou** [comment]
@Knand1861 Please don't bump multiple issues/pull requests for this. Instead, continue the discussion in https://github.com/godotengine/godot/issues/57180.

**Pull request merged**
